src/main/scala/mesosphere/marathon/core/launchqueue/impl/LaunchQueueActor.scala

```diff
@@ -198,16 +200,24 @@ private[impl] class LaunchQueueActor(
         case None => sender() ! None
       }
 
-    case Add(app, count) =>
+    case Add(runSpec, count) =>
       import context.dispatcher
 
       async {
-        val instances = 0.until(count).map { _ => Instance.Scheduled(app, Instance.Id.forRunSpec(app.id)) }
-        val start = await(instanceTracker.schedule(instances))
+        val existingReserved = await(instanceTracker.specInstances(runSpec.id))
+          .filter(_.isReserved)
+          .take(count)
+          .map(_.copy(state = InstanceState(Condition.Scheduled, Timestamp.now(), None, None), runSpecVersion = runSpec.version, unreachableStrategy = runSpec.unreachableStrategy))
+          .map(InstanceUpdateOperation.RelaunchReserved)
+        val instancesToSchedule = existingReserved.length.until(count).map { _ => Instance.Scheduled(runSpec, Instance.Id.forRunSpec(runSpec.id)) }
+        if (instancesToSchedule.nonEmpty) {
+          val scheduled = await(instanceTracker.schedule(instancesToSchedule))
+        }
+        val relaunched = await(Future.sequence(existingReserved.map(instanceTracker.process)))
```

Comment:
I asked myself the same question actually ... so in general I agree. But:
- Karsten refactored instance tracker to more CRUD interface quite recently so I would like to discuss with him
- but more importantly - as you said, this code will go away when goal state is introduced anyway - then only the `schedule` call will remain so it will actually be easier to remove and separate those two then

to sum up - I would vote for leaving as it is :)

Reply:
There is actually too much logic in this code. I think I mentioned before that the check should happen in the orchestrator. There shouldn't be many calls to `LaunchQueue.add`. Are there?